dbcmetrics/dbcm_logic/dbcuptimekumamaintenance.py

```diff
@@ -0,0 +1,112 @@
+import os
+import datetime
+from zoneinfo import ZoneInfo
+
+from dateutil.parser import parse
+import time
+import logging
+from prometheus_client import Gauge
+from uptime_kuma_api import UptimeKumaApi
+
+from dbcm_common.dbcmexception import DBCMException
+from dbcm_common.repeating_thread import RepeatingThread
+
+
+# A class to handle fetching maintenance windows from uptimekuma and exposing them as metrics.
+class UptimeKumaMaintenanceWindowThreading(object):
+  # Takes a configuration dictionary with keys like API URL, API key, timezone, etc.
+  def __init__(self, config: dict):
+    try:
+      # Initializes instance variables from the config.
+      file_configs = config["uptime_kuma_maintenance_metrics"]
+      self.API_URL = file_configs["api_url"]
+
+      # cannot use api key as the socket.io api needed for retrieval of planned maintenance does not support it, cf. https://github.com/louislam/uptime-kuma/issues/3625#issuecomment-1686760395
+      self.UPTIME_KUMA_USERNAME = os.getenv("UPTIME_KUMA_USERNAME")
+      self.UPTIME_KUMA_PASSWORD = os.getenv("UPTIME_KUMA_PASSWORD")
+
+      if self.UPTIME_KUMA_USERNAME is None or self.UPTIME_KUMA_PASSWORD is None:
+        logging.error("Missing uptime kuma credentials")
+        raise DBCMException
+
+      self.METRIC_REFRESH_INTERVAL = datetime.timedelta(seconds=file_configs["metric_refresh_interval_sec"])
+
+    except Exception as e:
+      logging.error("Missing or invalid configuration for Uptime Kuma API maintenance metrics.")
+      raise DBCMException from e
+
+    # stores maintenance windows.
+    self.windows = {}
+    # stores monitor ids and corresponding names
+    self.monitors = {}
+    # Prometheus gauge metric
+    self.metric = Gauge(
+      "uptime_kuma_maintenance_active",
+      "1 if any monitor is in a scheduled maintenance window, else 0",
+      ["monitor_id", "monitor_name"]
+    )
+
+    # Loads initial data, starts a background thread to refresh metrics periodically.
+    self.load_maintenance_windows()
+    self.last_loaded = time.time()
+
+    RepeatingThread(interval=self.METRIC_REFRESH_INTERVAL, name="Metrics Refresh", target=self.run)
+    logging.info(f"Uptime Kuma Maintenance Metrics Thread started. UTC Time: {datetime.datetime.now(datetime.UTC)}")
+
+  def run(self):
+    if (time.time() - self.last_loaded) > self.METRIC_REFRESH_INTERVAL.total_seconds():
```

Comment:
The interval here is the same as the interval for the RepeatingThread itself.
I think that condition is only useful if we have 2 different intervals for loading the maintenance windows and refreshing the metrics.